Sources/MapboxMaps/Foundation/Events/Payload/ResourceRequestPayload.swift

```diff
@@ -0,0 +1,76 @@
+import Foundation
+
+public struct ResourceRequestPayload {
+    public let dataSource: ResourceDataSource
+    public let request: ResourceRequest
+    public let response: ResourceResponse
+}
+
+extension ResourceRequestPayload: Decodable {
+    enum CodingKeys: String, CodingKey {
+        case dataSource = "data-source"
+        case request, response
+    }
+}
+
+public enum ResourceDataSource: String, Decodable {
+    case resourceLoader = "resource-loader", network, database, asset, fileSystem = "file-system"
+}
+
+public struct ResourceRequest: Decodable {
+    enum CodingKeys: String, CodingKey {
+        case url, kind, priority, loadingMethod = "loading-method"
+    }
+    public enum Kind: String, Decodable {
+        case unknown, style, source, tile, glyphs, spriteImage = "sprite-image", spriteJSON = "sprite-json", image
+    }
+    public enum Priority: String, Decodable {
+        case regular, low
+    }
+    public enum LoadingMethod: String, Decodable {
+        case cache, network
+    }
+
+    let url: String
+    let kind: Kind
+    let priority: Priority
+    let loadingMethod: [LoadingMethod]
+}
+
+public struct ResourceResponse: Decodable {
+    enum CodingKeys: String, CodingKey {
+        case noContent = "no-content"
+        case notModified = "not-modified"
+        case mustRevalidate = "must-revalidate"
+        case source, size, modified, expires, etag, error, cancelled
+    }
+    public enum Source: String, Decodable {
+        case network, cache, tileStore = "tile-store", localFile = "local-file"
+    }
+
+    public struct Error: Swift.Error, Decodable {
+        // swiftlint:disable:next nesting
+        public enum Reason: String, Decodable {
+            case success
+            case notFound = "not-found"
+            case server
+            case connection
+            case rateLimit = "rate-limit"
+            case inOfflineMode = "in-offline-mode"
```

Comment:
Yeah, not sure about an external library but we can think about implementing one, probably outside of the scope of this PR

Reply:
I believe it would be better to implement it as a part of this PR. There are tons of new `CodingKeys` that can be eliminated with appropriate `KeyDecodingStrategy`. It's also safer to run the "experimental" decoding strategy over some scope that you are very familiar with and have a total control.

The implementation also looks pretty straightforward. Something like:

```swift
decoder.keyDecodingStrategy = .custom { keys in
    func camelCased(_ string: String) -> String {
        return string.lowercased()
            .split(separator: "-")
            .enumerated()
            .map { $0.offset > 0 ? $0.element.capitalized : $0.element.lowercased() }
            .joined()
    }

    let lastKey = keys.last!.stringValue

    return AnyKey(stringValue: camelCased(lastKey))!
}
```

We also have to introduce `AnyKey` (intValue is required by CodingKey):

```swift
struct AnyKey: CodingKey {
    var stringValue: String
    var intValue: Int?

    init?(stringValue: String) {
        self.stringValue = stringValue
        self.intValue = nil
    }

    init?(intValue: Int) {
        self.stringValue = String(intValue)
        self.intValue = intValue
    }
}
```

Reply:
I can confirm that this implementation works 😎 
<img width="637" alt="image" src="https://user-images.githubusercontent.com/735178/173093213-317dfc4f-2ade-4cc7-bf08-a85f38423cf0.png">

Reply:
Can we still implement it as a part of this PR?